external-crates/move/move-compiler/src/diagnostics/mod.rs

```diff
@@ -340,6 +362,69 @@ macro_rules! diag {
     }};
 }
 
+impl WarningFilters {
+    pub fn new() -> Self {
+        Self::Empty
+    }
+
+    pub fn is_filtered(&self, diag: &Diagnostic) -> bool {
+        self.is_filtered_by_info(&diag.info)
+    }
+
+    fn is_filtered_by_info(&self, info: &DiagnosticInfo) -> bool {
+        match self {
+            WarningFilters::All => info.severity() == Severity::Warning,
+            WarningFilters::Specified { category, codes } => {
+                info.severity() == Severity::Warning
+                    && (category.contains(&info.category())
+                        || codes.contains(&(info.category(), info.code())))
+            }
+            WarningFilters::Empty => false,
+        }
+    }
+
+    pub fn union(&mut self, other: &WarningFilters) {
+        match (self, other) {
+            (s @ Self::Empty, _) => *s = other.clone(),
+            (_, Self::Empty) => (),
+            (Self::All, _) => (),
+            (s, Self::All) => *s = Self::All,
+            (
+                Self::Specified { category, codes },
+                Self::Specified {
+                    category: other_category,
+                    codes: other_codes,
+                },
+            ) => {
+                category.extend(other_category);
+                codes.extend(other_codes);
+            }
+        }
+    }
+
+    pub fn add(&mut self, filter: WarningFilter) {
+        match self {
+            WarningFilters::All => (),
+            WarningFilters::Empty => {
+                *self = WarningFilters::Specified {
+                    category: BTreeSet::new(),
+                    codes: BTreeSet::new(),
+                };
+                return self.add(filter);
+            }
+            WarningFilters::Specified { category, codes } => match filter {
+                WarningFilter::All => *self = WarningFilters::All,
+                WarningFilter::Category(cat) => {
+                    category.insert(cat);
+                }
+                WarningFilter::Code(cat, code) => {
```

Comment:
Similarly for this, if we're already filtering the category then we can just not add the category + code here as well I believe.